File: computingservices/DedupeServices/rstreamio/redisstreamwriter.py
import logging
from utils import redisstreamdb, notification_stream_key
from rstreamio.message.schemas.notification import NotificationPublishSchema
from datetime import datetime
from utils.loggingutils import log_event


logger = logging.getLogger(__name__)

class redisstreamwriter:

    rdb = redisstreamdb
    notificationstream = rdb.Stream(notification_stream_key)

    def sendnotification(self, message, error=False):
        try:
            notification_msg = NotificationPublishSchema()
            notification_msg.serviceid = "dedupe"
            notification_msg.errorflag = self.__booltostr(error)
            notification_msg.ministryrequestid = message.ministryrequestid
            notification_msg.createdby = message.createdby
            notification_msg.createdat = datetime.now().strftime("%m/%d/%Y, %H:%M:%S.%f")
            notification_msg.batch = message.batch
            #Additional execution parameters : Begin
            
            #Additional execution parameters : End
            msgid = self.notificationstream.add(notification_msg.__dict__, id="*")
            log_event(
                logger,
                logging.INFO,
                "notification_published",
                context=message,
                stream_id=msgid,
            )
        except RuntimeError as error:
            log_event(
                logger,
                logging.ERROR,
                "notification_publish_failed",
                context=message,
                exc_info=True,
            )

    def __booltostr(self, value):
        return "YES" if value == True else "NO"

File: computingservices/CompressionServices/rstreamio/redisstreamwriter.go
package rstreamio

import (
	"context"
	"log"
	"time"

	"compressionservices/models"

	"github.com/go-redis/redis/v8"
)

type NotificationPublishSchema struct {
	ServiceID         string `json:"serviceid"`
	ErrorFlag         string `json:"errorflag"`
	MinistryRequestID int    `json:"ministryrequestid"`
	CreatedBy         string `json:"createdby"`
	CreatedAt         string `json:"createdat"`
	Batch             string `json:"batch"`
}

// RedisStreamWriter publishes the existing flat notification message contract.
// Its Redis client and target stream are explicitly owned by the caller.
type RedisStreamWriter struct {
	rdb                *redis.Client
	notificationStream string
}

func NewRedisStreamWriter(redisClient *redis.Client, notificationStream string) *RedisStreamWriter {
	return &RedisStreamWriter{rdb: redisClient, notificationStream: notificationStream}
}

func (w *RedisStreamWriter) SendNotification(message *models.CompressionProducerMessage, errorFlag bool) {
	if w == nil || w.rdb == nil || message == nil {
		return
	}
	notificationMsg := map[string]interface{}{
		"serviceid":         "compression",
		"errorflag":         boolToStr(errorFlag),
		"ministryrequestid": message.MinistryRequestID,
		"createdby":         message.CreatedBy,
		"createdat":         time.Now().Format("2006-01-02 15:04:05.000"),
		"batch":             message.Batch,
	}
	if _, err := w.rdb.XAdd(context.Background(), &redis.XAddArgs{Stream: w.notificationStream, Values: notificationMsg}).Result(); err != nil {
		log.Print("Unable to write notification message")
	}
}

func boolToStr(value bool) string {
	if value {
		return "YES"
	}
	return "NO"
}
